### PR TITLE
Introduce Gradle wrapper

### PR DESCRIPTION
**Pull request by jasontedor**
We have agreed to introduce the Gradle wrapper to simplify workflows for developers, and managing infrastructure (e.g., CI, release builds, etc.) as well as consideration for the fact that other projects in our stack use Gradle and do not necessarily want to be tied to our Gradle version.

Relates #13744

### CODE REVIEW

**Review by nik9000 — APPROVED** [review]
I don't want the binary in the repo but I *really* don't want to wrangle with gradle upgrades and I like that we can rely on a particular version of gradle with this. This is an improvement overall so let's do it.

File: build.gradle [hunk 1]
```diff
@@ -402,6 +408,41 @@ task run(type: Run) {
...
+      def version = GradleVersion.version(wrapper.gradleVersion)
+      final URI uri =
+              locator.getDistributionFor(version, wrapper.distributionType.name().toLowerCase(Locale.ENGLISH))
+      uri.toURL().withInputStream { i ->
+        final byte[] buffer = new byte[1 << 20]
+        final MessageDigest sha256 = MessageDigest.getInstance("SHA-256")
+        int bytesRead
+        while ((bytesRead = i.read(buffer)) >= 0) {
+          if (bytesRead > 0) {
+            sha256.update(buffer, 0, bytesRead)
```
Comment (ywelsch) [inline build.gradle]: maybe simpler to just write 
```
String checksum = MessageDigest.getInstance("SHA-256").digest(uri.toURL().getBytes()).encodeHex().toString()
```

File: build.gradle [hunk 2]
```diff
@@ -402,6 +408,41 @@ task run(type: Run) {
...
+
+        wrapper.getPropertiesFile() << "distributionSha256Sum=${sha256.digest().encodeHex()}\n"
+      }
+    }
+  }
+
+  tasks.withType(Wrapper).each(wrapperClosure)
+
+  subprojects {
+    tasks.withType(Wrapper).each(wrapperClosure)
```
Comment (ywelsch) [inline build.gradle]: you can fold this call with the previous one into 
```
allprojects {
    tasks.withType(Wrapper).each(wrapperClosure)
}


```

File: build.gradle [hunk 3]
```diff
@@ -402,6 +408,41 @@ task run(type: Run) {
   impliesSubProjects = true
 }
 
+task wrapper(type: Wrapper)
+
+gradle.projectsEvaluated {
```
Comment (ywelsch) [inline build.gradle]: why is this required?

**Review by ywelsch — COMMENTED** [review]
Looks good to me. I've left two suggestions and one question.

File: build.gradle (hunk 1 above)
Comment (jasontedor) [reply build.gradle]: It is simpler, but I would prefer not to download an arbitrary number of bytes into memory in one go. This is why I chunk it.

**Review by ywelsch — APPROVED** [review]
